precompiles/erc20/approve_test.go

```diff
@@ -0,0 +1,562 @@
+	method := s.precompile.Methods[authorization.ApproveMethod]
+	amount := int64(100)
+
+	testcases := []struct {
+		name        string
+		malleate    func() []interface{}
+		postCheck   func()
+		expPass     bool
+		errContains string
+	}{
+		{
+			name:        "fail - empty args",
+			malleate:    func() []interface{} { return nil },
+			errContains: "invalid number of arguments",
+		},
+		{
+			name: "fail - invalid number of arguments",
+			malleate: func() []interface{} {
+				return []interface{}{
+					1, 2, 3,
+				}
+			},
+			errContains: "invalid number of arguments",
+		},
+		{
+			name: "fail - invalid address",
+			malleate: func() []interface{} {
+				return []interface{}{
+					"invalid address", big.NewInt(2),
+				}
+			},
+			errContains: "invalid address",
+		},
+		{
+			name: "fail - invalid amount",
+			malleate: func() []interface{} {
+				return []interface{}{
+					s.keyring.GetAddr(1), "invalid amount",
+				}
+			},
+			errContains: "invalid amount",
+		},
+		{
+			name: "fail - negative amount",
+			malleate: func() []interface{} {
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(-1),
+				}
+			},
+			errContains: "cannot approve non-positive values",
+		},
+		{
+			name: "pass - approve without existing authorization",
+			malleate: func() []interface{} {
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(amount),
+				}
+			},
+			expPass: true,
+			postCheck: func() {
+				s.requireSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetAccAddr(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, amount)),
+					[]string{},
+				)
+			},
+		},
+		{
+			name: "pass - approve with existing authorization",
+			malleate: func() []interface{} {
+				s.setupSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetPrivKey(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, 1)),
+				)
+
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(amount),
+				}
+			},
+			expPass: true,
+			postCheck: func() {
+				s.requireSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetAccAddr(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, amount)),
+					[]string{},
+				)
+			},
+		},
+		{
+			name: "pass - approve with existing authorization in different denomination",
+			malleate: func() []interface{} {
+				s.setupSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetPrivKey(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.bondDenom, 1)),
+				)
+
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(2 * amount),
+				}
+			},
+			expPass: true,
+			postCheck: func() {
+				s.requireSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetAccAddr(0),
+					// NOTE: The approval in the different denomination is overwritten by the
+					// approval for the passed token denomination.
+					//
+					// TODO: check if this behavior is the same for ERC20s? Or can there be separate
+					// approvals for different denominations?
```

Comment:
this is the expected behavior right?

Reply:
I think this is expected in the way Authz's work they are only constrained to a unique pair for granter/grantee.

Reply:
> I think this is expected in the way Authz's work they are only constrained to a unique pair for granter/grantee.

This is a bug. It shouldn't be replaced but added to the existing send authorization. 

Eg. 

1. I create an approval of 10 ATOM to address X
2. I create a second approval for 2 OSMO to address X

The send authorization should contain both denominations after the second approval